### ISSUE TITLE
Make FiltaQuilla compatible with Thunderbird 78

### ISSUE THREAD

**Issue by reagle** [RealRaven2000/FiltaQuilla#64]
I just got upgraded to TB 78.0, but FiltaQuilla is disabled.

**Comment by RealRaven2000** [comment]
We are working on upgrades for all my Add-ons, first my commercial ones (QuickFolders / quickFilters and SmartTemplate4). Once these are done and some new revenue comes in through license sales, I will start looking at my "free as in beer ones" - First Menu On top and Zombiekeys (as I have written them largely myself) and then later the inherited stuff from Kent (Filtaquilla, Tonequilla).

If I feel real adventurous I might pay somebody to turn QuickPasswords into a web extension - I miss it badly in Firefox, that's why I am using Waterfox for over a year now.

**Comment by RealRaven2000** [comment]
I have renamed this issue - all update efforts for 78 and any regressions specific to this version of Thunderbird will be logged here!

**Comment by bhafer** [comment]
Argh. Well, Thunderbird has now released 78.2.2 to the release channel. So FiltaQuilla broke. :(
Specifically, it's the change tag actions that I was using and now fail.

Just found this thread and wondering what things are looking like? Is there hope? Can I help?

**Comment by RealRaven2000** [comment]
> 
> 
> Argh. Well, Thunderbird has now released 78.2.2 to the release channel. So FiltaQuilla broke. :(
> Specifically, it's the change tag actions that I was using and now fail.
> 
> Just found this thread and wondering what things are looking like? Is there hope? Can I help?

Sure:
- if you're a developer, have a look at the source code and send me your github handle
- if you're a power user, spread the word and leave a review linking to this issue
- if you like to future proof you can donate to axel@songcraftstudios.com (topic: "FiltaQuilla Development") or buy a quickFilters / QuickFolders / SmartTemplate4 license. I am still fundraising to cover all development costs for upgrading, so far we have burned around $3,000.

Note: My COmpany is called Song Craft Studios because I am getting into music production full time. By the time that will pay my bills everything will be hopefully web (mail) extensions and much lower maintenance. I estimate I will probably have to sink around $20,000 into development and marketing to make it sustainable for all my Add-ons. I would honestly love to give FiltaQuilla to somebody else but there is nobody who has spare time and my (quickFilters) users rely on it so much.

**Comment by RealRaven2000** [comment]
I keep getting reviews that question why FiltaQuilla isn't compatible, so I made a more verbose reply (which also links here):

> what is it with addons regularly becoming incompatible with thunderbird? the extension works fine when it does, but messes up the filters when it is incompatible. would give full 5* if did not invalidate filters when incompatible. maybe set them to inactive somehow?

This comes from Mozilla's constant changes. They have a big one in preparation the switch to "mail extensions" which is what Firefox did a few years back when _they_ pulled the plug on legacy Add-ons. Thunderbird uses the Firefox code base (mozilla-central) as foundation, so they had to eventually follow suit.

This means Add-ons are only allowed special restricted APIs for carrying out their tasks - these are not complete!

Since we are not there yet (I would say not even 5% of what sophisticated Add-ons need in the Mail sphere) there is an interim step - packaging the Add-on as a "mail experiment". but this takes time and resources (and will still break functionality if not done correctly) - we have spent 250hours on conversion efforst for my Add-ons, but the paid licensed ones will have to be completed first before I can do it with the "free as in beer" ones. It is very hard to find developers with this knowledge!

If you are or know a developer who would like to get involved please feel free to leave a comment here:
https://github.com/RealRaven2000/FiltaQuilla/issues/64

If you want to help fund raising for converting FiltaQuilla consider buying a license for quickFilters Pro.

**Comment by jeremymeyers** [comment]
If there are specific API hooks that would make this conversion easier, I woudl be happy to submit a bugzilla thing about it.

**Comment by RealRaven2000** [comment]
> If there are specific API hooks that would make this conversion easier, I woudl be happy to submit a bugzilla thing about it.

You can try - I think it may be a bit early to handle it via API and might take months to get there - the quickest way to get this online would be by using a experiment API and the WindowListener (as I have done with my extensions QuickFolders / quickFilters and SmartTemplate4). 

There is a detailed cooking list on how to achieve this, here:
https://github.com/thundernest/addon-developer-support/wiki/Using-the-WindowListener-API-to-convert-a-Legacy-Overlay-WebExtension-into-a-MailExtension-for-Thunderbird-78

It's basically a step by step process. which I will attack once my commercial Add-ons have finally been published. Would you like to help?

**Comment by yakatz** [comment]
@RealRaven2000 I was just force-upgraded from 68 to 78. If you can still use help, I could probably spend some time on it, but I wanted to make sure nothing changed in the last 9 days before I start looking at it.

**Comment by RealRaven2000** [comment]
> If there are specific API hooks that would make this conversion easier, I would be happy to submit a bugzilla thing about it.

I am not sure if there are API hooks specific to filtering - the actual API submission process itself is still under discussion. I highly recommend subscribing to the [mailling list addons@discuss.thunderbird.net](https://thunderbird.topicbox.com/groups/addons) (there is a thread "[Defining the process to discuss and propose future APIs](https://thunderbird.topicbox.com/groups/addons/Tac2edf01a4a3db12/defining-the-process-to-discuss-and-propose-future-apis)"). For the meantime, I am hoping to use the more "hacky" method of using the WindowLoader to inject the code and still use legacy methods (XPCOM) for accessing all functionality directly.

**Comment by RealRaven2000** [comment]
> @RealRaven2000 I was just force-upgraded from 68 to 78. If you can still use help, I could probably spend some time on it, but I wanted to make sure nothing changed in the last 9 days before I start looking at it.

Absolutely - feel free to fork the project and submit PRs if you change anything. Please use this new ESR78 branch as starting point:

https://github.com/RealRaven2000/FiltaQuilla/tree/ESR78

**Comment by RealRaven2000** [comment]
I just checked in conversion step1 for Tb78 in the ESR78 branch. THis "skeleton" loads but gives one error:

window[this.uniqueRandomID].onLoad is not a function

**Comment by decadent** [comment]
Shooting in the dark here, but afaik event handlers are written in lowercase for the olde HTML events, i.e. ‘`onload`’.

Additionally, if handlers are added via `addEventListener`, then afaik they can't be invoked through the `onsomething` calls—instead the event needs to be fired [with dispathEvent](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent). However, it's also weird to have non-standard arguments on event handlers, so I'm not entirely sure what's going on in the code.

**Comment by RealRaven2000** [comment]
> Shooting in the dark here, but afaik event handler properties are written in lowercase for the olde HTML events, i.e. ‘`onload`’.
> 
> Additionally, if handlers are added via `addEventListener`, then afaik they can't be invoked through the `onsomething` calls—instead the event needs to be fired [with dispatchEvent](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/dispatchEvent). However, it's also weird to have non-standard arguments on event handlers, so I'm not entirely sure what's going on in the code.

I was able to fix that one - I had simply not added the onLoad and onUnload functions to the window scripts yet. Now I am super stuck at the bundle strings - for some reason I can registerd my string bundle at the location "chrome://filtaquilla/locale/filtaquilla.properties" but it always crashes out when I call getStringFromName(). Do I need a physical "chrome" directory?

here is my chrome registration list:
```
  messenger.WindowListener.registerChromeUrl([ 
      ["content",  "filtaquilla",           "content/"],
      ["resource", "filtaquilla",           "skin/"],
      ["resource", "filtaquilla",           "content/"],
      ["locale",   "filtaquilla", "en",     "locale/en-US/"],
      ["locale",   "filtaquilla", "sv",     "locale/sv-SE/"],
      ["locale",   "filtaquilla", "de",     "locale/de/"],
      ["locale",   "filtaquilla", "nl",     "locale/nl/"]
    ]
  );  
```

I checked in the WindowListener mplementation and apparently they are added correctly.

**Comment by RealRaven2000** [comment]
> ```
>   const bundleService = Cc["@mozilla.org/intl/stringbundle;1"].getService(Ci.nsIStringBundleService),
>         filtaquillaStrings = bundleService.createBundle("chrome://filtaquilla/locale/filtaquilla.properties");
>  ..
> .. some object:
>   name: filtaquillaStrings.GetStringFromName("filtaquilla.applyIncomingFilters")
> ```
> 
> the last line above throws `NS_ERROR_UNEXPECTED`

I managed to fix that one, turns out you cannot have 2 different resources with the same name. Once I renamed the skin to 

`["resource", "filtaquilla-skin",           "skin/"],`

and adjusted my chrome urls accordingly it worked.

**Comment by RealRaven2000** [comment]
This afternoon I managed to build an early "alpha version", which can be downloaded from here:

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0pre12

My main issues

- there are still some bindings left that need to be converted to classes. I haven't done this in a while, and there was probably a reason why I didn't do the remaining ones. We will see
- I had to convert the version code to something asynchronous, which took me most of the day today. But at least the version number now shows up in the options dialog correctly. I also used the preferences code I had hacked with @TbSync for connecting the preference elements with the preferences in the options dialog, looks like I got that part working again
- know issue: if you load the Add-on with debugger multiple times, the added conditions in filter editor are duplicated. I am not quite sure what causes this but I guess something goes wrong when the customElements are registered (there is no way to de-register an existing customElement, so that may be a bit of a gap in the definitions) 
- pretty sure we have to get rid of the action to run JavaScript. I don't see that this will pass the review on ATN, and it's probably a real edge case anyway,.

**Comment by kgraefe** [comment]
>     * pretty sure we have to get rid of the action to run JavaScript. I don't see that this will pass the review on ATN, and it's probably a real edge case anyway,.

FTR, this is the only action I am interested in: https://github.com/kgraefe/thunderbird-autotag 

However, when I switch to TB78 it might be time to turn this into its own addon.

**Comment by RealRaven2000** [comment]
New version with a quickfix for #74 

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0pre13

Other than that, same comments as 2 comments above apply.

**Comment by tneele** [comment]
> pretty sure we have to get rid of the action to run JavaScript. I don't see that this will pass the review on ATN, and it's probably a real edge case anyway,.

Just chipping in as well: I use the 'run Javascript' action to modify the subjects of messages. The IT department of my organisation has decided to put '[EXT]' in the subject of every email coming from an external sender, which is very annoying. FiltaQuilla was the only addon I could find to help me fix this.

**Comment by aetreyu** [comment]
Will the Javascript "Search Term" still be supported? Please say yes :-)

**Comment by RealRaven2000** [comment]
> Will the Javascript "Search Term" still be supported? Please say yes :-)

I am hoping, but THunderbird may not permit it. I am still trying to figure out how to get a textbox showing for any of the regular expressions - I failed ti get it converted from xbl bindings in Novmber 2019 already, and have wasted 2 solid days this week not getting anywhere. Not bad for a "free as in beer" Add-on.

**Comment by lgs11** [comment]
I am so happy that my regex filters are working again with 3.0pre13. I've had them for a while and became dependent on them. I don't think I'll need to edit them anytime soon. Thanks!

**Comment by RealRaven2000** [comment]
> I am so happy that my regex filters are working again with 3.0pre13. I've had them for a while and became dependent on them. I don't think I'll need to edit them anytime soon. Thanks!

If they are already working, that's good - let's hope they will be passed through ATN though. They can be very pissy about running scripts from an Add-on.

**Comment by p91paul** [comment]
@RealRaven2000 are you aware that Mozilla has paid people to help developers with addons transitioning? see this comment on another extension for example, you might try contacting the guy if you're stuck somewhere https://github.com/wangvisual/autoarchive/issues/118#issuecomment-695915225

**Comment by RealRaven2000** [comment]
yes. i have been bringing this up only 150 times over the last few days
both on the addon devs and the thunderbird development chat on element /
matrix.  John (@jobisoft) already helped with the conversion of my Add-ons
Quickfolders and SmartTemplates (whereby he also improved his experimental
apis WindowListener and LegacyPrefs) and i got paid help from @opto
<buecher@optosolar.com>  for all the ground up issues.

i got some valuable hints from Alex one off the core developers today so i
have written some code to inject my textboxes into the search-value-custom
boxes without using bindings

**Comment by elhennig** [comment]
Is there a problem with the filter depending on the folder name the message is in? With TB68 this is still working and I rely on that filter feature

**Comment by RealRaven2000** [comment]
> Is there a problem with the filter depending on the folder name the message is in? With TB68 this is still working and I rely on that filter feature

I don't know yet, I cannot even edit any condition that has a textbox, so I can't test it. If we do not have the textbox to edit the condition, we have nothing. I am on the third day trying to fix the textbox (or rather make it appear) - since bindings have been taken away I need to inject it somehow, which is extremely hard.

**Comment by wonzling** [comment]
Hello, i want to use only one action: "Detach Attachments To". But it seems that this not work. But when i try to use: "Save Attachments" To this works. I'm using the Thunderbird 78 with 3.0pre13 Version.
Sorry for the english ....

**Comment by RealRaven2000** [comment]
Here is my first attempt at fixing the missing textboxes for filter conditions, sans bindings:

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0pre16

this should at least give you guys a chance to edit regular expressions for subject, attachment, header as well as conditions cased on `bcc` and folder name. The next step will be looking at the tag specific conditions. The last one will be the JavaScript condition.

**Comment by wonzling** [comment]
I'm sorry, but the "Detach Attachtment To" function stíll not works ...

**Comment by RealRaven2000** [comment]
> I'm sorry, but the "Detach Attachtment To" function stíll not works ...

don't be sorry, this is exactly what this thread is for! Let's just get FiltaQuilla compatible with THundebird 78 first (it's super hard) and once it is released and stuff like that still doesn't work, we can open a fresh github issue. Attachments are difficult, as they often make Thunderbird hang / crash (especially if you are using IMAP, because you need to download the full mail, and IMAP is terrible with that). Hopefully there will be some stable API functions from the Tb core team for this soon...

**Comment by wonzling** [comment]
OK thank you for your quick response

**Comment by PhilPsychology** [comment]
As a user, I very much appreciate the efforts and collaboration being shown. Whatever the process required to get Filtaquilla compatible, I support that and can see it is very challenging.

Like wonzling, I am very interested ultimately in being able to resume using "detach all" on more than one email at a time. I do not expect it to be perfect and I accept that very large attachments and groups of emails may cause hanging or crashing, but regardless, some degree of this functionality returning is much appreciated.

**Comment by RealRaven2000** [comment]
I am still "trolling" the Tb Add-on Developers chat fishing for information. It seems that custom Elements (which Kent's original FiltaQuilla extension is using) will not be allowed in future, instead we should be implementing the filter editor extensions with experimental APIs; these have to eventually be integrated into Thunderbird Core:
![image](https://user-images.githubusercontent.com/6314674/99253697-1b229500-2809-11eb-9cf7-81381a3ef9d7.png)

I am just putting this here so I can at least quote one Thunderbird developer in case they start to question my code next year - in case experimental Add-ons are forbidden or relegated to "Thunderbird beta versions only" - as it was done with Firefox years ago.

**Comment by RealRaven2000** [comment]
Another prerelease, this one supports tags.

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0pre19

![image](https://user-images.githubusercontent.com/6314674/99262161-87a39100-2815-11eb-9a2c-5578567beb93.png)

(important caveat - you need to open the filter editor with the tag condition in it so you can edit it - I am still missing working code for when the user changes the search condition to "Tag of Thread Messages" manually)

**Comment by RealRaven2000** [comment]
> > ```
> > * pretty sure we have to get rid of the action to run JavaScript. I don't see that this will pass the review on ATN, and it's probably a real edge case anyway,.
> > ```
> 
> FTR, this is the only action I am interested in: https://github.com/kgraefe/thunderbird-autotag
> 
> However, when I switch to TB78 it might be time to turn this into its own addon.

I fixed the JavaScript Action part so if you would like to test, here is a version:

[filtaquilla-3.0pre21.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5547049/filtaquilla-3.0pre21.zip)

I still need to fix the JavaScript for search condition, not quite sure how that works

**Comment by El-Pazzo** [comment]
Hi @RealRaven2000,
could it be that the command "add sender to address book" is wrongly matched to "add sender to group?

regards,
Pascal

**Comment by LittleAlf** [comment]
Just a feedback regarding the following FiltaQuilla settings being used by me (filtaquilla-3.0pre16.zip):

**1) Save attachments to [Anhänge speichern in]**
Works as in TB68 :) / _small regression:_ When editing the settings and hitting the 'folder' symbol, the current folder is not selected, but rather its parent folder (maybe this is on purpose)

**2) Run Program [Programm ausführen]**
Works as in TB68 :) / _regression_: executed command line is not visible; cannot be properly edited, i.e. to enter a parameter

Nevertheless, both work already. **And I thank you very, very much for your work.**

Best regards
LittleAlf

**Comment by RealRaven2000** [comment]
> Just a feedback regarding the following FiltaQuilla settings being used by me (filtaquilla-3.0pre16.zip):
> 
> **1) Save attachments to [Anhänge speichern in]**
> Works as in TB68 :) / _small regression:_ When editing the settings and hitting the 'folder' symbol, the current folder is not selected, but rather its parent folder (maybe this is on purpose)

you have to be "in the folder" and click "select folder" (not click on a folder) - so it is by design
> 
> **2) Run Program [Programm ausführen]**
> Works as in TB68 :) / _regression_: executed command line is not visible; cannot be properly edited, i.e. to enter a parameter

Ok, for Run Program, here is another trial version - it hadn't been properly patched to use `<html:input>` instead of `<textbox>`:

[filtaquilla-3.0pre22.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5549858/filtaquilla-3.0pre22.zip)

Looks like this in Win7:
![image](https://user-images.githubusercontent.com/6314674/99310207-90ff1e80-2852-11eb-967f-2e717c51cff4.png)

**Comment by El-Pazzo** [comment]
Hi @RealRaven2000 ,

regarding the issue with the address book, please find a screenshot attached to see what I mean:

![adressbuchproblem](https://user-images.githubusercontent.com/47013454/99369987-6ac29880-28bd-11eb-86c0-502399efeaab.png)

"Absenderadress zu Gruppe hinzufügen" was formerly "Absenderadresse zu Adressbuch hinzufügen", the problem is that the dropdown only shows "Spam" but not the other address books that are connected with the account. This filter for example should use "Persönliches Adressbuch".

PS: am using your latest release (pre22)

regards,
Pascal

**Comment by RealRaven2000** [comment]
> Hi @RealRaven2000 ,
> 
> regarding the issue with the address book, please find a screenshot attached to see what I mean:
> 
> "Absenderadress zu Gruppe hinzufügen" was formerly "Absenderadresse zu Adressbuch hinzufügen", the problem is that the dropdown only shows "Spam" but not the other address books that are connected with the account. This filter for example should use "Persönliches Adressbuch".
> 

So the English term Kent used was "Add Sender to Address List" - I may have wrongly assumed that he meant a group instead of the address book. IN my test instance I am only able to add "New List" under the "Personal AB"

![image](https://user-images.githubusercontent.com/6314674/99384135-e2e18c00-28c6-11eb-8bc1-91d91d3fbe36.png)

I think these are the lists that the command refers to. I didn't find a way to create a new Address book entry at the top level,

After creating a filter action in a filter, I am getting something that look like an empty dropdown list:
![image](https://user-images.githubusercontent.com/6314674/99384300-1de3bf80-28c7-11eb-9734-d585f74bd87a.png)

- but it is also possible that the list is truncated by the parent element for some reason. (there is a gray bar underneath that indicates there are hidden items there). So I need to investigate whether / how this is filled at first.

**Comment by RealRaven2000** [comment]
This is the code that generates the menuitems - apparently there is no internal "label" element created, so they appear to be empty vessels in the DOM:

```
    addDirectories(aDirEnum, aMenupopup) {
      while (aDirEnum.hasMoreElements()) {
        let dir = aDirEnum.getNext();
        if (dir instanceof Ci.nsIAbDirectory) {
          // get children
          let newMenuItem = document.createElement('menuitem'),
              displayLabel = dir.dirName;
          if (dir.isMailList)
            displayLabel = "  " + displayLabel;
          newMenuItem.setAttribute('label', displayLabel);
          newMenuItem.setAttribute('value', dir.URI);
          newMenuItem.classList.add('menuitem-iconic');
          aMenupopup.appendChild(newMenuItem);
          // recursive add of child mailing lists
          let childNodes = dir.childNodes;
          if (childNodes && childNodes.hasMoreElements())
            this.addDirectories(childNodes, aMenupopup);
        }
      }
    }
```

**Comment by RealRaven2000** [comment]
Ok, did a bit of work on that - turns out I have to use createXULElement (createElement only makes a html without the added xul "goodness"). I also added a bit of styling:

![image](https://user-images.githubusercontent.com/6314674/99395066-cef25600-28d7-11eb-8155-7323a1de1e1d.png)

Apparently, at the moment, the value is only stored if you select a list item, none of the Address Book items work. Also it should probably be more hierarchical. but it's a start. Test version:

[filtaquilla-3.0pre23.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5553718/filtaquilla-3.0pre23.zip)

**Comment by El-Pazzo** [comment]
Thanks @RealRaven2000 ,

Now it works for the selection. Only the German translation needs to be updated to "Absender zu Adressbuch hinzufügen".

- If I now open an existing filter with such an address list set in the action part, the wrong (in my case Spam instead of Persönliches Adressbuch) is selected but it is correctly saved in the filter file. An address book set in the conditions part is correctly displayed.

**Comment by RealRaven2000** [comment]
> Now it works for the selection. Only the German translation needs to be updated to "Absender zu Adressbuch hinzufügen".
> 

Just on this label - it cannot do that I think the correct term is "Verteilerliste" - it can only add addresses to the lists with the "Person" icon not to the AB root "Addressbuch". At least that's the original implementation as far as I understand it
![image](https://user-images.githubusercontent.com/6314674/99418939-73ce5c80-28f3-11eb-9189-418f1dc42f21.png)

**Comment by El-Pazzo** [comment]
The problem is not that we need to be able to add to Verteilerliste (a sublist of Adressbuch) but to Adressbuch itself. This option is not given right now.
Adding to a Verteilerliste is of course also interestig but I guess most users will need "add to address book".

So if I get it right, right now the addon tries to add an address to a Verteilerliste (which does not exist, thus throws an error) instead of adding it to the address book.

**Comment by El-Pazzo** [comment]
@RealRaven2000 
I just realized that in order to really use the "Verteilerliste" (mailing list) for the filters, we would need to be able to use them as conditions too, but there is no way to select them. So in conditions you only have address books for selection an in actions you only have "Verteilerlisten" (mailing lists) as selection option. ;)

**Comment by RealRaven2000** [comment]
> @RealRaven2000
> I just realized that in order to really use the "Verteilerliste" (mailing list) for the filters, we would need to be able to use them as conditions too, 

how so? The "add contact..." action should definitely guard against duplicates automatically.

**Comment by El-Pazzo** [comment]
@RealRaven2000 
if it was included for the address book, yes. But it is not included, or does condition "if included in address book" also check mailing lists? (Which would of course make sense.)

Right now Action part has: Add to mailing list
it misses: Add to address book

Right now Condition part has:  … is included/is not included in > address book
it misses: … is included/is not included in > mailing list

mailing list and address book are 2 completely different lists, so if we cannot add addresses to address book in filters, the condition to check address book does not fully work if it relies on a filter that would add such addresses there, the same for mailing lists.

On the other hand, what if people do not want to work with mailing lists, they'll then not be able to use your filter for such purposes or be forced to create mailing lists just for this filter action. It would at least be good if people could chose the action they want to use. Either "add to mailing list" or "add to address book".

For the condition part it would also be good for those working with mailing lists, to be able to check whether an address is part of a mailing list (but not the entire address book).

**Comment by El-Pazzo** [comment]
@RealRaven2000 
@RealRaven2000 

I think what confused me here is that condition and action have different names in German. 

sorry, I mixed up 2 threads and forgot that address books already are included here. The issue that still persists though, is that although I set an address book correctly the address book is not "locked" when reopening the filter. I added "persönliches Adressbuch" to an action but when reopening that filter it shows "Spam" as the set address book. The filter itself also throws an error on all the filters using such a set action, as if the set address book would not exist or something like that. So to me it seemed that due to the different naming in German it tried to add to a mailing list instead of the set address book.

**Comment by El-Pazzo** [comment]
I just checked the mail filter file and realized that apparently the action "add to mailing list" (Absenderadresse zu Gruppe hinzufügen) does not get updated at all. All other actions and conditions do get updated though.

**Comment by RealRaven2000** [comment]
I managed to "fix" the editor (at least for changing that list action from the drop down) - I am leaving the test of the actual filter action to you, ok?

[filtaquilla-3.0pre24.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5560104/filtaquilla-3.0pre24.zip)

It took a bit of "finnicking around" so that the menulist value would be copied up to the correct parent element. But you can select an addressbook now, not just an address list. Not sure what would happen if you set the action to add the address to "Collected Addresses". btw how do you create a new AB called "Spam"? I honestly don't know how to create additional address books from the UI.

**Comment by El-Pazzo** [comment]
@RealRaven2000 

the action value now gets saved but for some reason the addSender action still throws an error. I tried with mailing lists and the address book directly. It does not work. If I remove that action the filters work fine.

**Comment by RealRaven2000** [comment]
> @RealRaven2000
> 
> the action value now gets saved but for some reason the addSender action still throws an error. I tried with mailing lists and the address book directly. It does not work. If I remove that action the filters work fine.

I looked at the code, apparently the function MailServices.headerParser.parseHeadersWithArray(..) doesn't exist anymore. I will write some shim code for it...

**Comment by RealRaven2000** [comment]
(prerelease 25) ...from a quick test, it seems to add to the AB nodes as well - but doesn't guard against duplicates there, so still some more work for me to do.

**Comment by El-Pazzo** [comment]
I still get the error message. For some reason the error is not logged in the log file, all other actions are logged though.

**Comment by RealRaven2000** [comment]
...so from what I can see the biggest problem is still this major one:

- the filtaquilla conditions are not functional when the user selects one from the dropdown - they only work when the filter is loaded with them already existing

![image](https://user-images.githubusercontent.com/6314674/99538494-f400ca80-29a4-11eb-92dd-15f2e9f3bee5.png)

**Comment by RealRaven2000** [comment]
> ...so from what I can see the biggest problem is still this major one:
> 
> * the filtaquilla conditions are not functional when the user selects one from the dropdown - they only work when the filter is loaded with them already existing
> 

that stuff is very difficult! The Filter Editor provides one <search-value-custom..> element which by rights should be removed and re-added with the correct type. Very hard to debug and code, I have burned another 4 hours on it and come to the conclusion that it is even more complicated as I thought. All because bindings were removed!

**Comment by LittleAlf** [comment]
Final feedback regarding the following FiltaQuilla filter actions (filtaquilla-3.0pre22.zip):

- Save attachments to
- Run Program

**Both work as expected and identical to FiltaQuilla in TB68.**

Thank you very much.

LittleAlf

**Comment by El-Pazzo** [comment]
Thanks a lot for the work so far @RealRaven2000 ;)

**Comment by RealRaven2000** [comment]
Here is another quick alpha version. I think we have sorted at least the editor for regex conditions and the tag selector. Still to do: Javascript editor and then a bunch of testing.

[filtaquilla-3.0pre32.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5578714/filtaquilla-3.0pre32.zip)

**Comment by RealRaven2000** [comment]
> Will the Javascript "Search Term" still be supported? Please say yes :-)

this is the next one   i am ging to tackle. i just repaired the Edit box / tag selectors for the other conditions when the user changes to the FiltaQuilla conditions (tag header / regex header etc) so that the correct custom elements are injected into the custom search value box. i will follow the same patterns for the JS box and i already converted the JS editor dialog to xhtml, so we should have working UI for testing JS conditions tomorrow.

**Comment by RealRaven2000** [comment]
I fixed the JavaScript editor control for the JavaScript condition - here is a new test version:

[filtaquilla-3.0pre39.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5579504/filtaquilla-3.0pre39.zip)

Still has the ugly button, but we can now edit and save script.
![image](https://user-images.githubusercontent.com/6314674/99904352-3aff0080-2cc2-11eb-94b8-ab849d298656.png)

Any ideas how to test this? What's a good real world example? @aetreyu

**Comment by RealRaven2000** [comment]
> > pretty sure we have to get rid of the action to run JavaScript. I don't see that this will pass the review on ATN, and it's probably a real edge case anyway,.
> 
> Just chipping in as well: I use the 'run Javascript' action to modify the subjects of messages. The IT department of my organisation has decided to put '[EXT]' in the subject of every email coming from an external sender, which is very annoying. FiltaQuilla was the only addon I could find to help me fix this.

Please test pre39 above!

**Comment by tneele** [comment]
I tried the latest test version in Thunderbird 78.5.0 on Ubuntu 20.04. The Javascript action works well for me, I didn't run into any issues. For reference, here's the script I'm using:
```
for (let index = 0; index < msgHdrs.length; index++)
{
    let hdr = msgHdrs.queryElementAt(index, Ci.nsIMsgDBHdr);
    hdr.subject = hdr.subject.replace("[EXT] ", "");
}
```
Thanks a bunch for all your work on FiltaQuilla! I sent you a small donation.

**Comment by aetreyu** [comment]
@RealRaven2000

Here's an (edited) example that could be used for testing (e.g. you could star every message coming from or to a gmail.com address).

Use case: prior to editing, the first condition was used to match addresses on a specific company domain. The second two conditions were to match a few exception addresses outside the company domain. All e-mail from or two matching addresses gets moved to a specific folder.

`//anonymous function (to prevent variable and function conflicts)
(function(){
	var pattern;
	var match=false;

	pattern=/gmail\.com/i;
	if(pattern.test(message.author)||pattern.test(message.recipients)||pattern.test(message.ccList)||pattern.test(message.bccList))
	{
		match=true;
	}

	pattern=/bob@example\.com/i;
	if(pattern.test(message.author)||pattern.test(message.recipients)||pattern.test(message.ccList)||pattern.test(message.bccList))
	{
		match=true;
	}

	pattern=/jill@isp\.net/i;
	if(pattern.test(message.author)||pattern.test(message.recipients)||pattern.test(message.ccList)||pattern.test(message.bccList))
	{
		match=true;
	}

	if(match)
	{
		return true;
	}
	else
	{
		return false;
	}
})();`

**Comment by micaelmedia** [comment]
Hi!

We would love to see the print mail-function from this script back again. Can pay or donate to what needed if that helps.

**Comment by ondohotola** [comment]
I need to recognise Attachments by name, RegEx doesn't seem to work yet, but some wildcards would also work for me. But maybe I can circumvent this via the Subject line which is parseable.

greetings, el

**Comment by caviidae** [comment]
The 'Folder name' filter was the most important thing to me. How hard would that be to bring back?

**Comment by elhennig** [comment]
> The 'Folder name' filter was the most important thing to me. How hard would that be to bring back?

For me this works since 3.0pre19

**Comment by CardBook** [comment]
@RealRaven2000 : how to make a donation for your work ?

**Comment by RealRaven2000** [comment]
> @RealRaven2000 : how to make a donation for your work ?

that would be via the quickFilters donations page:
https://quickfilters.quickfolders.org/donate.html#donate
I am trying to keep FiltaQuilla as a "free as in beer" product...

**Comment by RealRaven2000** [comment]
Just made a new official prerelease, so people can download the latest here:

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0pre40

**Comment by RealRaven2000** [comment]
Fixed the "print" action - this currently uses the msgPrintEngine window https://searchfox.org/comm-esr78/source/mailnews/base/content/msgPrintEngine.xhtml which was renamed from msgPrintEngine.xul

[filtaquilla-3.0pre41.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5610920/filtaquilla-3.0pre41.zip)

**Comment by micaelmedia** [comment]
Thanks for fixing the print-function! Donated on the quickFilters page.

**Comment by RealRaven2000** [comment]
I just published a new prerelease:

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0pre44

this fixes printing, move later, copy as read and the debug settings dialog.

**Comment by RealRaven2000** [comment]
Just a heads up on some related work - [quickFilters 5.0](https://addons.thunderbird.net/en-US/thunderbird/addon/quickfilters/versions/?page=1#version-5.0) has just been reviewed and accepted on ATN, so please install it if you haven't got it yet! We need larger numbers to support this (and other) Add-on projects that are related to filtering emails.

**Comment by El-Pazzo** [comment]
Hi @RealRaven2000,

today I saw that filter condition "Betreff Vergleichen (Regex…"  does not show the text field if you open an existing filter. It's greyed out so you cannot edit it.

**Comment by RealRaven2000** [comment]
> Hi @RealRaven2000,
> 
> today I saw that filter condition "Betreff Vergleichen (Regex…" does not show the text field if you open an existing filter. It's greyed out so you cannot edit it.

it works for me, are you using the latest beta (3.0pre44) ?

![image](https://user-images.githubusercontent.com/6314674/100886841-31d51700-34ac-11eb-92a5-8c11a01f9af4.png)

**Comment by El-Pazzo** [comment]
Yes, installed it today. If I add a new condition it works, but as soon as I reopen it for editing, it's greyed out.

**Comment by El-Pazzo** [comment]
Ok, I just saw that for another rule I created today, it works fine, so I tried to recreate one of the rules that don't work from scratch but again, it does not work.

**Comment by RealRaven2000** [comment]
> Ok, I just saw that for another rule I created today, it works fine, so I tried to recreate one of the rules that don't work from scratch but again, it does not work.

Can you make me a screenshot so I can see which line it is please?

**Comment by El-Pazzo** [comment]
![filtaquilla issue](https://user-images.githubusercontent.com/47013454/101024284-2a794080-3574-11eb-85ca-4437c9f0c017.png)
here we go. 

and here the spam rule that goes with it:

name="Spam 2"
enabled="yes"
type="17"
action="JunkScore"
actionValue="100"
action="Delete"
condition="AND (from,isn't in ab,jsaddrbook://abook.sqlite) AND (filtaquilla@mesquilla.com#subjectRegex,matches,\"((Arabic|Czech|Dutch|English|French|Italian|Japanese|Korean|Polish|Slovak|Spanish|Swedish|Vietnamese)(\s)?(<(-)?>|to|-)(\s)?(Arabic|Chinese|Czech|Dutch|French|English|Greek|Hindi|Italian|Japanese|Korean|Polish|Portuguese|Spanish|Turkish|Vietnamese))|(Excellent quality translations)|Profile of \b.*\b Translator|(Bilingual|Freelance) Translator\")"

**Comment by RealRaven2000** [comment]
```
name="Spam 2"
enabled="yes"
type="17"
action="JunkScore"
actionValue="100"
action="Delete"
condition="AND (from,isn't in ab,jsaddrbook://abook.sqlite) AND ([filtaquilla@mesquilla.com](mailto:filtaquilla@mesquilla.com)#subjectRegex,matches,"((Arabic|Czech|Dutch|English|French|Italian|Japanese|Korean|Polish|Slovak|Spanish|Swedish|Vietnamese)(\s)?(<(-)?>|to|-)(\s)?(Arabic|Chinese|Czech|Dutch|French|English|Greek|Hindi|Italian|Japanese|Korean|Polish|Portuguese|Spanish|Turkish|Vietnamese))|(Excellent quality translations)|Profile of \b.*\b Translator|(Bilingual|Freelance) Translator")"

```
that one wouldn't load for me (After adding to mine, Thunderbird just threw away my complete msgRules.dat and created a new empty one. I can create it manually though - or you could export with quickFilters?
![image](https://user-images.githubusercontent.com/6314674/101042221-d6bb2700-3574-11eb-859e-4aa3f7c6a7df.png)

**Comment by RealRaven2000** [comment]
Hmnm. this was working before now it's broken again. probbaly a regression between 40 and 44. or maybe a side effect of quickFilters, I need to see...

**Comment by RealRaven2000** [comment]
It seems to be triggered by the actual content of your regular expression - when the dialog loads the text field is missing. Need to debug this.

**Comment by RealRaven2000** [comment]
One more test - using the same string in an "ordinary" text field (subject) doesn't seem to affect Thunderbird in the same way:

![image](https://user-images.githubusercontent.com/6314674/101046455-f6ece500-3578-11eb-9b96-773050293184.png)

**Comment by RealRaven2000** [comment]
Ok I think I found out what is happening - I am injecting the value as an attribute in xul:
![image](https://user-images.githubusercontent.com/6314674/101098040-2bcd5c00-35ba-11eb-828f-72376d51fa5c.png)

I think the < and > characters break the xul parsing process. I will see if I can code a workaround that sets the value directly.

**Comment by RealRaven2000** [comment]
Can you try this quick-fix version?

[filtaquilla-3.0pre45.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5639476/filtaquilla-3.0pre45.zip)

Instead of trying to inject the value within the xul fragment, I have changed the code (for both the textbox and also the javascript values) as follows:

```
  let textbox = window.MozXULElement.parseXULToFragment(
    ` <html:input class="search-value-textbox flexinput fq-textbox" inherits="disabled" 
      onchange="this.parentNode.setAttribute('value', this.value); this.parentNode.value=this.value;"> 
      </html:input>`
  );
  let inputElement = es.appendChild(textbox);
  inputElement.value = es.getAttribute("value");
  es.classList.add("flexelementcontainer");

```

**Comment by RealRaven2000** [comment]
Well - it should have worked but now the value isn't copied up to the parent element and so is never saved in the filter. NO idea why this happens now (unless it was broken from the beginning). This is stuff is complex...

**Comment by RealRaven2000** [comment]
ok, I found out that appendChild in this case does not return (as specified by DOM standard) the html element but another document fragment. so I need to get the lastChild instead, should be easy enough to fix:
```
  let textbox = window.MozXULElement.parseXULToFragment(
    ` <html:input class="search-value-textbox flexinput fq-textbox" inherits="disabled" 
      onchange="this.parentNode.setAttribute('value', this.value); this.parentNode.value=this.value;"> 
      </html:input>`
  );
  es.appendChild(textbox);
  es.lastChild.value = es.getAttribute("value");
  es.classList.add("flexelementcontainer");
```

And here is a final working version that accepts your regular expression
`((Arabic|Czech|Dutch|English|French|Italian|Japanese|Korean|Polish|Slovak|Spanish|Swedish|Vietnamese)(\s)?(<(-)?>|to|-)(\s)?(Arabic|Chinese|Czech|Dutch|French|English|Greek|Hindi|Italian|Japanese|Korean|Polish|Portuguese|Spanish|Turkish|Vietnamese))|(Excellent quality translations)|Profile of \b.*\b Translator|(Bilingual|Freelance) Translator`

[filtaquilla-3.0pre46.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5639588/filtaquilla-3.0pre46.zip)

**Comment by El-Pazzo** [comment]
Yes, that one works.

**Comment by RealRaven2000** [comment]
The latest release for Thunderbird 78 (as of today) is here:

https://github.com/RealRaven2000/FiltaQuilla/releases/tag/3.0

I uploaded it today but shortly after got a message that 2.1.1 (which was reviewed earlier this week) does not update on ATN, so I did an emergency release 2.1.2 - which disabled my 3.0 release. Apparently we cannot review 2 things at the same time, the ATN version only sees the last upload. Which means I have to bump to 3.0.1 and release once 2.1.2 has been finally reviewed.

**Comment by bhafer** [comment]
I downloaded and installed v3.0.1, and I want to say that it is working wonderfully for managing tags with filters.
Fantastic! Thank you :)

**Comment by aetreyu** [comment]
@RealRaven2000

thanks so much for your hard work! i've had to block updates to thunderbird because i depend on your addon....

could you clarify which features, if any, are lost in 3.0 compared to 2.1.2?

thanks again! :-)

**Comment by RealRaven2000** [comment]
> @RealRaven2000
> 
> thanks so much for your hard work! i've had to block updates to thunderbird because i depend on your addon....
> 
> could you clarify which features, if any, are lost in 3.0 compared to 2.1.2?
> 
> thanks again! :-)

I don't have the full list but theoretically they should work the same as in Thunderbird 68. the ones I tested in particular:

### Actions
* print mail
* detach attachment
* save attachment
* run a program
* launch file  
* Move Later
(File related: there are some issues with UNC paths, which I will fix shortly, see issue #79)

### Search Terms 
* regular expressions (Subject Regex Match) - if others regex functions don't work please raise a separate issue
* bcc
* Tag of Thread Messages / Head (I did some fixes there, but feel free to do some deep testing)

Which features were you interested in particularly? If you find something that doesn't work, just make a separate issue so I can pin it to fix first!

**Comment by aetreyu** [comment]
My filter search terms and actions are working :-)

Tested:

- "Folder Name" "is"
- "Status" "is" "New"
- "From, To, Cc or Bcc" "contains"
- "From" "isn't in my address book"
- "Javascript" "matches"

- "Copy Message To"
- "Move Message To"
- "Prepend to Subject"
- "Mark As Read"
- "Add Star"
- "Stop Filter Execution"
- "Save message as file"
- "Run Program" (with and without MESSAGEID parameter)
- "Delete Message"

**Comment by RealRaven2000** [comment]
> My filter search terms and actions are working :-)
> 

I assume you upgraded and this means for Thundebrird 78? That would be great news!

**Comment by Alaindeg** [comment]
Is it here the place to ask some help to use FiltaQuilla ?

If not, could you give me a link for that ?

Thank you

**Comment by Boomatang** [comment]
@RealRaven2000 I would like to say thank you for you work. FiltaQuilla updated this morning and now my zero inbox  is working as a it should.  :smiley: :partying_face:

**Comment by RealRaven2000** [comment]
> Is it here the place to ask some help to use FiltaQuilla ?
> 

![image](https://user-images.githubusercontent.com/6314674/102100236-1c9fa700-3e21-11eb-9ff0-0a0b3c4574ef.png)

this links to the support site: https://quickfilters.quickfolders.org/filtaquilla.html

I usually have a contact me link at the bottom of every page (it's under my quickFilters website because both do related stuff). It is possible that the link is missing on that particular page - which I need to rectify. It will be on any of the quickFilters pages in the set (just click any of the links on the top menu)

**Comment by RealRaven2000** [comment]
Closing this issue as the official version 3.0.1 - now supports Thunderbird 78:
https://addons.thunderbird.net/en-US/thunderbird/addon/filtaquilla/
